nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -89,94 +119,370 @@ contract WorkLock {
+                }
+            }
+        }
+
+        require(ethToTokens(minBid) > maxAllowableLockedTokens,
+            "At least one of bidders has allowable bid");
+
+        // final bids adjustment (only for bonus part)
+        // (min_whale_bid * token_supply - max_stake * eth_supply) / (token_supply - max_stake * n_whales)
+        uint256 maxBonusTokens = maxAllowableLockedTokens - minAllowableLockedTokens;
+        uint256 minBonusETH = minBid - minAllowedBid;
+        uint256 bonusTokenSupply = tokenSupply - bidders.length * minAllowableLockedTokens;
+        uint256 refundETH = minBonusETH.mul(bonusTokenSupply)
+                                .sub(maxBonusTokens.mul(bonusETHSupply))
+                                .divCeil(bonusTokenSupply - maxBonusTokens.mul(length));
+        uint256 resultBid = minBid.sub(refundETH);
+        bonusETHSupply -= length * refundETH;
+        for (uint256 i = 0; i < length; i++) {
+            address bidder = _biddersForRefund[i];
+            WorkInfo storage info = workInfo[bidder];
+            refunds[i] += refundETH;
+            info.depositedETH = resultBid;
+        }
+
+        // reset verification
+        nextBidderToCheck = 0;
+
+        // save a refund
+        for (uint256 i = 0; i < length; i++) {
+            address bidder = _biddersForRefund[i];
+            compensation[bidder] += refunds[i];
+            emit ForceRefund(msg.sender, bidder, refunds[i]);
+        }
+
+    }
+
+    /**
+    * @notice Withdraw compensation after force refund
+    */
+    function withdrawCompensation() external {
+        uint256 refund = compensation[msg.sender];
+        require(refund > 0, "There is no compensation");
+        compensation[msg.sender] = 0;
+        msg.sender.sendValue(refund);
+        emit CompensationWithdrawn(msg.sender, refund);
+    }
+
+    /**
+    * @notice Check that the claimed tokens are within `maxAllowableLockedTokens` for all participants,
+    * starting from the last point `nextBidderToCheck`
+    * @dev Method stops working when the remaining gas is less than `_gasToSaveState`
+    * and saves the state in `nextBidderToCheck`.
+    * If all bidders have been checked then `nextBidderToCheck` will be equal to the length of the bidders array
+    */
+    function verifyBiddingCorrectness(uint256 _gasToSaveState) external afterCancellationWindow returns (uint256) {
+        require(nextBidderToCheck != bidders.length, "Bidders have already been checked");
+
+        // all participants bid with the same minimum amount of eth
+        uint256 index = nextBidderToCheck;
+        if (bonusETHSupply == 0) {
+            require(tokenSupply / bidders.length <= maxAllowableLockedTokens, "Not enough bidders");
+            index = bidders.length;
+        }
+
+        uint256 maxBonusTokens = maxAllowableLockedTokens - minAllowableLockedTokens;
+        uint256 bonusTokenSupply = tokenSupply - bidders.length * minAllowableLockedTokens;
```

Comment:
Indeed interesting, this will make this method very cheap (for this case) but will slightly increase gas cost for other cases. I guess if it's realistic to achieve such number of bidders let's add this